### PR TITLE
Add register_partition, unregister_partition procedures

### PR DESCRIPTION
**Pull request by wendigo**
Hive: Add partition manipulation procedures:

* ``system.register_partition(schema_name, table_name, location, partition_columns, partition_values)``

    Registers existing location as a partition in metastore for specified table.

* ``system.unregister_partition(schema_name, table_name, partition_columns, partition_values)``

    Unregisters given, existing partition in metastore for specified table.
    The referenced partition data location is not deleted.

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java [hunk 1]
```diff
@@ -97,14 +97,12 @@ public void createEmptyPartition(ConnectorSession session, String schema, String
...
     {
         TransactionalMetadata hiveMetadata = hiveMetadataFactory.get();
+        HiveTableHandle tableHandle = (HiveTableHandle)hiveMetadata.getTableHandle(session, new SchemaTableName(schemaName, tableName));
 
-        ConnectorTableHandle tableHandle = hiveMetadata.getTableHandle(session, new SchemaTableName(schemaName, tableName));
-        HiveInsertTableHandle hiveInsertTableHandle = (HiveInsertTableHandle) hiveMetadata.beginInsert(session, tableHandle);
-
-        List<String> actualPartitionColumnNames = hiveInsertTableHandle.getInputColumns().stream()
-                .filter(HiveColumnHandle::isPartitionKey)
+        List<String> actualPartitionColumnNames = tableHandle.getPartitionColumns().stream()
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java]: > Refactor create_empty_partition

you can call this commit sth like
```
Verify columns before beginInsert() in create_empty_partition
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java [hunk 2]
```diff
@@ -216,7 +216,8 @@ private static void dropPartitions(
                     session,
                     table.getDatabaseName(),
                     table.getTableName(),
-                    extractPartitionValues(name));
+                    buildPartitionObject(session, table, name),
+                    true);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java]: it actually should be `false`, as we think the partition does not exist

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 3]
```diff
@@ -745,26 +745,31 @@ public synchronized void addPartition(
...
     {
         setShared();
         Map<List<String>, Action<PartitionAndMore>> partitionActionsOfTable = partitionActions.computeIfAbsent(new SchemaTableName(databaseName, tableName), k -> new HashMap<>());
-        Action<PartitionAndMore> oldPartitionAction = partitionActionsOfTable.get(partitionValues);
+        Action<PartitionAndMore> oldPartitionAction = partitionActionsOfTable.get(partition.getValues());
         if (oldPartitionAction == null) {
             HdfsContext hdfsContext = new HdfsContext(session, databaseName, tableName);
             HiveIdentity identity = new HiveIdentity(session);
-            partitionActionsOfTable.put(partitionValues, new Action<>(ActionType.DROP, null, hdfsContext, identity));
+            PartitionStatistics stats = PartitionStatistics.empty();
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: it seems unreasonable to create stats object to _drop_ a partiton
maybe we shouldn't be reusing `PartitionAndMore` here and use a different (new?) type instead?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 4]
```diff
@@ -1175,6 +1175,41 @@ public void testCreateEmptyNonBucketedPartition()
...
+    public void testUnregisterRegisterPartition()
+    {
+        String tableName = "test_register_partition_for_table";
+        assertUpdate("" +
+                "CREATE TABLE " + tableName + " (" +
+                "  dummy_col bigint," +
+                "  part varchar)" +
+                "WITH (" +
+                "  format = 'ORC', " +
+                "  partitioned_by = ARRAY[ 'part' ] " +
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: `ORC` is default, so you can put whole `WITH` clause on one line
(also, we would write `['part']` rather than `[ 'part' ]`)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 3 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: Initially I’ve done that but in the end it required a lot of changes in different places.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        }
+    }
+
+    private void doRegisterPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
+    {
+        HiveIdentity identity = new HiveIdentity(session);
+        HdfsContext hdfsContext = new HdfsContext(session, schemaName, tableName);
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+
+        Path partitionLocation = new Path(location);
```
Comment (aalbu) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: In Hive's `ALTER TABLE ... ADD PARTITION`, `LOCATION` is optional.  I found that to be a great usability factor - I didn't want to worry about typos in specifying the path.  Unfortunately, procedures don't support optional arguments, so you would have to pass in a `null`, which is kind of ugly, but we do in other procedures.  Have we considered not asking the user to specify a location?  We pretty much have the logic for that implemented in `SyncPartitionMetadataProcedure`.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHivePartitionProcedures.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+    private Long getTableCount(String tableName)
+    {
+        QueryResult countResult = query("SELECT COUNT(*) FROM " + tableName);
+        return (Long) countResult.row(0).get(0);
+    }
+
+    private Set<String> getPartitionValues(String tableName)
+    {
+        return query("SELECT col FROM " + tableName).rows().stream().map(row -> row.get(0)).map(it -> (String) it).collect(Collectors.toSet());
```
Comment (aalbu) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHivePartitionProcedures.java]: You could use `.map(String.class::cast)`.  Same thing below.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHivePartitionProcedures.java [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+        hdfsClient.createDirectory(path);
+        HiveDataSource dataSource = createResourceDataSource(tableName, "io/prestosql/tests/hive/data/single_int_column/data.textfile");
+        hdfsDataSourceWriter.ensureDataOnHdfs(path, dataSource);
+    }
+
+    private static void createPartitionedTable(String tableName)
+    {
+        query("DROP TABLE IF EXISTS " + tableName);
+
+        query("CREATE TABLE " + tableName + " (val int, col varchar) WITH (format = 'TEXTFILE', partitioned_by = ARRAY[ 'col'])");
```
Comment (aalbu) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHivePartitionProcedures.java]: Formatting: extraneous space in `ARRAY[ 'col']`.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 8]
```diff
@@ -883,11 +883,24 @@ private void writeRoleGrantsFile(Set<RoleGrant> roleGrants)
...
     }
 
+    private boolean isValidPartition(Path tableMetadataDirectory, String partitionName)
+    {
+        Path metadata = new Path(new Path(tableMetadataDirectory, partitionName), PRESTO_SCHEMA_FILE_NAME);
+
+        try {
+            return metadataFileSystem.getFileStatus(metadata).isFile();
+        }
+        catch (Exception e) {
```
Comment (aalbu) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java]: Perhaps we should be more specific and just catch `FileNotFoundException` (which according to the Javadoc is what is thrown if the path does not exist)?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 9]
```diff
@@ -2169,8 +2169,7 @@ public ActionType getType()
 
         public T getData()
         {
-            checkState(type != ActionType.DROP);
-            return data;
+            return requireNonNull(data, "data is null");
```
Comment (aalbu) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: This pattern seems to go against our guidelines.  How about making `data` an `Optional<T>`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+        if (table.getPartitionColumns().isEmpty()) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Table is not partitioned: " + schemaTableName);
+        }
+
+        List<String> actualPartitionColumnNames = table.getPartitionColumns().stream()
+                .map(Column::getName)
+                .collect(toImmutableList());
+
+        if (!Objects.equals(partitionColumnNames, actualPartitionColumnNames)) {
```
Comment (aalbu) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: I think the comparison should be case insensitive.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/UnregisterPartitionProcedure.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+        if (table.getPartitionColumns().isEmpty()) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Table is not partitioned: " + schemaTableName);
+        }
+
+        List<String> actualPartitionColumnNames = table.getPartitionColumns().stream()
+                .map(Column::getName)
+                .collect(toImmutableList());
+
+        if (!Objects.equals(partitionColumnNames, actualPartitionColumnNames)) {
```
Comment (aalbu) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/UnregisterPartitionProcedure.java]: Case insensitive?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java (hunk 2 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java]: it changes behaviour (it was true before) but makes sense. We don't want data here to be dropped :)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 3 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: Ok. I've changed it other way.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java (hunk 5 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: Makes sense. I've made PR for optional procedure arguments: https://github.com/prestosql/presto/pull/2706

Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: When it got merge I'll make a follow up PR for making location optional (with default)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java (hunk 8 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java]: Actually it's IOException. Done

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 9 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: Refactored

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java (hunk 10 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: I'll make a follow-up PR changing this in every procedure (create empty partition as well)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/UnregisterPartitionProcedure.java (hunk 11 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/UnregisterPartitionProcedure.java]: I'll make a follow-up PR changing this in every procedure (create empty partition as well)

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 12]
```diff
@@ -738,6 +738,15 @@ Procedures
     ``create_empty_partition``). A null value for the ``partition_values`` argument indicates that stats
     should be dropped for the entire table.
 
+* ``system.register_partition(schema_name, table_name, location, partition_columns, partition_values)``
+
+    Registers existing location as a partition in metastore for specified table.
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/hive.rst]:
```suggestion
    Registers existing location as a new partition in the metastore for the specified table.
```

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 13]
```diff
@@ -738,6 +738,15 @@ Procedures
...
+
+* ``system.unregister_partition(schema_name, table_name, partition_columns, partition_values)``
+
+    Unregisters given, existing partition in metastore for specified table.
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/hive.rst]:
```suggestion
    Unregisters given, existing partition in the metastore for the specified table.
```

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 14]
```diff
@@ -738,6 +738,15 @@ Procedures
...
+    The referenced partition data location is not deleted.
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/hive.rst]:
```suggestion
    The partition data is not deleted.
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java [hunk 15]
```diff
@@ -97,14 +96,14 @@ public void createEmptyPartition(ConnectorSession session, String schema, String
     private void doCreateEmptyPartition(ConnectorSession session, String schemaName, String tableName, List<Object> partitionColumnNames, List<Object> partitionValues)
     {
         TransactionalMetadata hiveMetadata = hiveMetadataFactory.get();
-        HiveTableHandle tableHandle = (HiveTableHandle)hiveMetadata.getTableHandle(session, new SchemaTableName(schemaName, tableName));
+        HiveTableHandle tableHandle = (HiveTableHandle) hiveMetadata.getTableHandle(session, new SchemaTableName(schemaName, tableName));
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java]: This belongs to the previous commit

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java [hunk 16]
```diff
@@ -24,7 +24,6 @@
 import io.prestosql.spi.PrestoException;
 import io.prestosql.spi.classloader.ThreadContextClassLoader;
 import io.prestosql.spi.connector.ConnectorSession;
-import io.prestosql.spi.connector.ConnectorTableHandle;
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java]: This belongs to the previous commit

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java [hunk 17]
```diff
@@ -97,14 +96,14 @@ public void createEmptyPartition(ConnectorSession session, String schema, String
...
-        HiveTableHandle tableHandle = (HiveTableHandle)hiveMetadata.getTableHandle(session, new SchemaTableName(schemaName, tableName));
+        HiveTableHandle tableHandle = (HiveTableHandle) hiveMetadata.getTableHandle(session, new SchemaTableName(schemaName, tableName));
 
         List<String> actualPartitionColumnNames = tableHandle.getPartitionColumns().stream()
                 .map(HiveColumnHandle::getName)
                 .collect(toImmutableList());
 
         if (!Objects.equals(partitionColumnNames, actualPartitionColumnNames)) {
-            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "input partition column names doesn't match actual partition column names");
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Provided partition column names don't match actual partition column names");
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CreateEmptyPartitionProcedure.java]: We avoid contractions in messages for clarity (i'm aware previous code used them)

```suggestion
            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Provided partition column names do not match actual partition column names");
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    public void registerPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            doRegisterPartition(session, schemaName, tableName, location, partitionColumnNames, partitionValues);
+        }
+    }
+
+    private void doRegisterPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: What happens if any of arguments is NULL?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+        HdfsContext hdfsContext = new HdfsContext(session, schemaName, tableName);
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+
+        Path partitionLocation = new Path(location);
+
+        Table table = metastore.getTable(identity, schemaName, tableName)
+                .orElseThrow(() -> new TableNotFoundException(schemaTableName));
+
+        if (table.getTableType().equals(TableType.VIRTUAL_VIEW.name())) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Table is a view: " + schemaTableName);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: We should filter out `MATERIALIZED_VIEW` as well.

@electrum any idea why `CreateEmptyPartitionProcedure` doesn't have such a check?
did we consider this redundant over metastore validation?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+        if (table.getPartitionColumns().isEmpty()) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Table is not partitioned: " + schemaTableName);
+        }
+
+        List<String> actualPartitionColumnNames = table.getPartitionColumns().stream()
+                .map(Column::getName)
+                .collect(toImmutableList());
+
+        if (!Objects.equals(partitionColumnNames, actualPartitionColumnNames)) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, format("Provided partition column names don't match actual partition column names: %s", actualPartitionColumnNames));
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]:
```suggestion
            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, format("Provided partition column names do not match actual partition column names: %s", actualPartitionColumnNames));
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        List<String> partitionStringValues = partitionValues.stream()
+                .map(String.class::cast)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: Can't we declare `List<Object> partitionColumnNames, List<Object> partitionValues` as `List<String>`s?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+                .collect(toImmutableList());
+
+        String partitionName = FileUtils.makePartName(actualPartitionColumnNames, partitionStringValues);
+        Optional<Partition> partition = metastore.getPartition(new HiveIdentity(session), schemaName, tableName, partitionStringValues);
+
+        if (partition.isPresent()) {
+            throw new PrestoException(ALREADY_EXISTS, format("Partition %s already registered at location %s", partitionName, partition.get().getStorage().getLocation()));
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]:
```suggestion
            throw new PrestoException(ALREADY_EXISTS, format("Partition [%s] is already registered with location %s", partitionName, partition.get().getStorage().getLocation()));
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        if (!HiveWriteUtils.pathExists(hdfsContext, hdfsEnvironment, partitionLocation)) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Partition location does not exist: " + partitionLocation.toString());
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]:
```suggestion
            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Partition location does not exist: " + partitionLocation);
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+        metastore.commit();
+    }
+
+    private static Partition buildPartitionObject(ConnectorSession session, Table table, String partitionName, Path location)
+    {
+        return Partition.builder()
+                .setDatabaseName(table.getDatabaseName())
+                .setTableName(table.getTableName())
+                .setColumns(table.getDataColumns())
+                .setValues(extractPartitionValues(partitionName))
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: We know the values, we don't need to extract them.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 25]
```diff
@@ -745,15 +751,19 @@ public synchronized void addPartition(
         }
     }
 
-    public synchronized void dropPartition(ConnectorSession session, String databaseName, String tableName, List<String> partitionValues)
+    public synchronized void dropPartition(ConnectorSession session, String databaseName, String tableName, List<String> partitionValues, boolean dropData)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: There are many changes here which are related to adding `dropData` flag.
This makes it hard to follow.

Are you positive they are all necessary?

Anyway, it would be best to separate them into a prep commit (which should be easy -- generally everything except the two new procedures). This would allow giving the changes a name that explains the change _here_.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 26]
```diff
@@ -2144,20 +2152,14 @@ private void checkHoldsLock()
...
         private final HiveIdentity identity;
 
         public Action(ActionType type, T data, HdfsContext hdfsContext, HiveIdentity identity)
         {
             this.type = requireNonNull(type, "type is null");
-            if (type == ActionType.DROP) {
-                checkArgument(data == null, "data is not null");
-            }
-            else {
-                requireNonNull(data, "data is null");
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: - We had validation that `data != null` (unless it's DROP action). We no longer have it.

- I'm not a fan of nullable fields, but it appears that change `T data` to `Optional<T> data` is causing a lot of changes here, making the actual behavioral change invisible. I would back this out from this PR.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 27]
```diff
@@ -2286,41 +2287,90 @@ public String toString()
...
     {
         private final HiveIdentity identity;
-        private final Partition partition;
-        private final Path currentLocation;
-        private final Optional<List<String>> fileNames;
-        private final PartitionStatistics statistics;
-        private final PartitionStatistics statisticsUpdate;
 
-        public PartitionAndMore(HiveIdentity identity, Partition partition, Path currentLocation, Optional<List<String>> fileNames, PartitionStatistics statistics, PartitionStatistics statisticsUpdate)
+        public PartitionAndMore(HiveIdentity identity)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: Now the `PartitionAndMore` class doesn't have partition and doesn't have anything more.
I see it's used for inheritance, but i'd inline it.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 28]
```diff
@@ -1175,6 +1175,40 @@ public void testCreateEmptyNonBucketedPartition()
...
+                "  dummy_col bigint," +
+                "  part varchar)" +
+                "WITH (" +
+                "  partitioned_by = ARRAY['part'] " +
+                ")");
+
+        assertQuery(format("SELECT count(*) FROM \"%s$partitions\"", tableName), "SELECT 0");
+
+        assertUpdate(format("INSERT INTO %s (dummy_col, part) VALUES (1, 'first'), (2, 'second'), (3, 'third')", tableName), 3);
+        MaterializedResult results = getQueryRunner().execute(getSession(), "SELECT \"$path\" FROM " + tableName + " ORDER BY \"$path\" ASC").toTestTypes();
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]:
```suggestion
        MaterializedResult paths = getQueryRunner().execute(getSession(), "SELECT \"$path\" FROM " + tableName + " ORDER BY \"$path\" ASC").toTestTypes();
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 26 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: I've followed review comments and then introduced Optional<T> on data

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java (hunk 18 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: It's not possible - it's validated when the procedure is called (https://github.com/prestosql/presto/blob/62ecc58e8667d58ac5907a9f8bc72540eb8eea18/presto-main/src/main/java/io/prestosql/execution/CallTask.java#L142).

That's why I've added https://github.com/prestosql/presto/pull/2706

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 9 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: @aalbu see my comment https://github.com/prestosql/presto/pull/2692#discussion_r374928793:

> * We had validation that `data != null` (unless it's DROP action). We no longer have it.
> * I'm not a fan of nullable fields, but it appears that change `T data` to `Optional<T> data` is causing a lot of changes here, making the actual behavioral change invisible. I would back this out from this PR.

if we want to keep it, this needs to be a separate commit (this isn't really just few lines of cleanup), and we should retain existing validation.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 26 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: answered there (https://github.com/prestosql/presto/pull/2692#discussion_r373675285)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java (hunk 21 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: Yes we can (checked it)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 25 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: I've separated commits.

I'm positive that this changes are required.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 27 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: It's needed because partition manipulation actions are held together in ```@GuardedBy("this")
    private final Map<SchemaTableName, Map<List<String>, Action<PartitionAndMore>>> partitionActions = new HashMap<>()``` so either I will introduce interface, common super class or change partitionsActions to use ```Action<?>```

**Comment by wendigo** [comment]
Failed due to: ```2020-02-05T12:03:40.6380315Z java.lang.RuntimeException: java.util.concurrent.ExecutionException: org.testcontainers.containers.ContainerLaunchException: Container startup failed```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 9 above)
Comment (aalbu) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: My intent wasn't to suggest removing validation, just setting the value to `Optional.empty()` instead of null. For many callers, returning an `Optional` seemed a good fit, but I didn't realize the footprint of the change.  Sorry.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 26 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: Reverted most of the changes here

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 29]
```diff
@@ -883,11 +883,22 @@ private void writeRoleGrantsFile(Set<RoleGrant> roleGrants)
 
         List<String> partitionNames = partitions.stream()
                 .map(partitionValues -> makePartitionName(table.getPartitionColumns(), ImmutableList.copyOf(partitionValues)))
+                .filter(partitionName -> isValidPartition(table, partitionName))
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java]: I'd squash `Fix FileHiveMetastore to check if partition ...` commit with the next commit.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java (hunk 5 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java]: Let's gather more feedback before adding more flexibility. We would want to add the right options that are commonly needed or cannot be done otherwise.

**Comment by wendigo** [comment]
Failed due to https://github.com/prestosql/presto/issues/2616

**Comment by findepi** [comment]
```
2020-02-06T12:10:50.7843778Z ptl-tests           | 2020-02-06 17:55:50 INFO: FAILURE     /    io.prestosql.tests.hive.TestHivePartitionProcedures.testRegisterPartitionFromAnyLocation (Groups: smoke, hive_partitioning) took 2.8 seconds
2020-02-06T12:10:50.7844184Z ptl-tests           | 2020-02-06 17:55:50 SEVERE: Failure cause:
2020-02-06T12:10:50.7844495Z ptl-tests           | java.lang.AssertionError: 
2020-02-06T12:10:50.7844887Z ptl-tests           | Actual and expected have the same elements but not in the same order, at index 0 actual element was:
2020-02-06T12:10:50.7845177Z ptl-tests           |   <42>
2020-02-06T12:10:50.7845462Z ptl-tests           | whereas expected element was:
2020-02-06T12:10:50.7845733Z ptl-tests           |   <1>
2020-02-06T12:10:50.7845973Z ptl-tests           | 
2020-02-06T12:10:50.7846379Z ptl-tests           | 	at io.prestosql.tests.hive.TestHivePartitionProcedures.testRegisterPartitionFromAnyLocation(TestHivePartitionProcedures.java:196)
2020-02-06T12:10:50.7846729Z ptl-tests           | 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

`org.assertj.core.api.IterableAssert#containsExactly` checks order too, even for `assertThat(set)`

**Comment by findepi** [comment]
Merged as 783cb687ef333f62e2c00a079290086894dfd450, thanks!
